libr/anal/sign.c

```diff
@@ -774,6 +766,146 @@ R_API bool r_sign_add_anal(RAnal *a, const char *name, ut64 size, const ut8 *byt
 	return retval;
 }
 
+static RSignGraph *r_sign_fcn_graph(RAnalFunction *fcn) {
+	r_return_val_if_fail (fcn, false);
+	RSignGraph *graph = R_NEW0 (RSignGraph);
+	if (!graph) {
+		return NULL;
+	}
+	graph->cc = r_anal_function_complexity (fcn),
+	graph->nbbs = r_list_length (fcn->bbs);
+	graph->edges = r_anal_function_count_edges (fcn, &graph->ebbs);
+	graph->bbsum = r_anal_function_realsize (fcn);
+	return graph;
+}
+
+static RSignBytes *r_sign_fcn_bytes(RAnal *a, RAnalFunction *fcn) {
+	r_return_val_if_fail (a && fcn, false);
+	RCore *core = a->coreb.core;
+	int maxsz = r_config_get_i (core->config, "zign.maxsz");
+	int fcnlen = r_anal_function_realsize (fcn);
+	int size = R_MIN (core->io->addrbytes * fcnlen, maxsz);
+
+	// hopefully not necissary, but just in case
+	if (size <= 0) {
+		eprintf ("error: Negative size for bytes signature\n");
+		return NULL;
+	}
+
+	// get bytes part of signature
+	ut8 *bytes = malloc (size);
+	if (!bytes) {
+		return NULL;
+	}
+	if (!r_io_is_valid_offset (core->io, fcn->addr, 0)) {
+		eprintf ("error: cannot read at 0x%08" PFMT64x "\n", fcn->addr);
+		free (bytes);
+		return NULL;
+	}
+	(void)r_io_read_at (core->io, fcn->addr, bytes, size);
```

Comment:
I saw a comment in `r_io_read_at` that says it is deprecated. So I will also switch to using `r_io_nread_at`.

Reply:
actually r_io_read_at should be replaced with r_io_nread_at, iirc nread returns the amount of bytes read, and r_io_read_at does the same but returns bool depending if it can read all the bytes or none.

So we can make RIO.read_at call RIO.nread_at and save some lines in r_io..

* r_io_read_at : 251 refs
* r_io_nread_at: 6 refs

So it's clear that in most places in r2 it's checking for a complete read on pa/va depending on io.va, this probably requires further discussion with @condret and others to decide the way to go. I see useful behaviours in both apis, not sure if it should be deprecated or just overriden by nread